src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -370,15 +370,67 @@ def deleteBlocks(self):
         """
         _deleteBlocks_
         Find deletable blocks, then decide if to delete based on:
-        Is there an active subscription for dataset or block ?
-          If yes => set deleted=2
-          If no => next check
         Has transfer to all destinations finished ?
-          If yes => request block deletion, approve request, set deleted=1
+          If yes => Delete rules associated with the block, set deleted=1
           If no => do nothing (check again next cycle)
         """
-        # FIXME: figure out the proper logic for rule block deletion
-        logging.info("Starting deleteBlocks methods --> IMPLEMENT-ME!!!")
+        # Get list of blocks that can be deleted
+        blockDict = self.findDeletableBlocks.execute(transaction=False)
+
+        if not blockDict:
+            logging.info("No candidate blocks found for rule deletion")
+            return
+        else:
+            logging.info("Found %d candidate blocks for rule deletion", len(blockDict))
+
+        blocksToDelete = []
+        containerDict = {}
+        # Populate containerDict, assigning each block to its correspondant container
+        for blockName in blockDict:
+            container = blockDict[blockName]['dataset']
+            if container not in containerDict:
+                # Set of sites to wich the container is needs to be transferred
+                sites = blockDict[blockName]['sites']
+                containerDict[container] = {'blocks': [], 'rse': sites}
+            containerDict[container]['blocks'].append(blockName)
+
+        for contName in containerDict:
+            cont = containerDict[contName]
+
+            # Checks if the container is not requested in any sites.
+            # This should never be triggered, but better safe than sorry
+            if not cont['rse']:
+                logging.warning("No rules for container: %s. It's blocks won't be deleted.", contName)
+                continue
+
+            # Get RSE in which each block is available
+            availableRSEs = self.rucio.getReplicaInfoForBlocks(block=cont['blocks'])
+
+            for blockRSEs in availableRSEs:
+                # If block is available at every RSE its container needs to be transferred, the block can be deleted
+                if cont['rse'].issubset(blockRSEs['replica']):
+                    blocksToDelete.append(blockRSEs['name'])
+
+        # Delete agent created rules locking the block
+        binds = []
+        logging.info("Going to delete %d blocks", len(blocksToDelete))
+        for block in blocksToDelete:
+            rules = self.rucio.listDataRules(block, scope=self.scope, account=self.rucioAcct)
+            if not rules:
```

Comment:
This might be a problem in the future, because the `listDataRules` API catches Rucio exceptions and return an empty list. So it could be that: a) there are no real rules for that block anymore; or b) the query fails and we proceed marking the block as deleted.

The only option I see would be to stop catching that exception in the Rucio wrapper and let it propagate to whoever is making the call. But that would be a single wrapper API with a different behaviour... so I'm afraid I don't have any good suggestion for now.

Reply:
I think the best will be to change `listDataRules` to raise an `WMRucioException` exception in case that call fails. You catch it here then and handle it accordingly.